### ISSUE TITLE
Ability to reset the inner sub map

### ISSUE THREAD

**Issue by marioroy** [greg7mdp/parallel-hashmap#238]
Hi, @greg7mdp 

I tested variable length keys. It turns out the [llil4map.cc](https://gist.github.com/marioroy/862fa2fc6aa3b6f523f7a6ef9dd8d157) demonstration consumes the most memory consumption. The reason is unable to clear or reset the inner sub map. Without this ability, memory consumption peaks 18.4 GB. Otherwise, 15.1 GB with `reset_inner` capability.

Here, we get ready to insert the key,value pairs into a vector.

```c++
         propvec.resize(map.size());
         std::vector<vec_str_int_type::iterator> I; I.resize(map.subcnt());
         I[0] = propvec.begin();
         size_t prev_num_keys;

         for (size_t i = 0; i < map.subcnt(); ++i) {
            map.with_submap(i, [&](const map_str_int_type::EmbeddedSet& set) {
               if (i == 0) {
                  prev_num_keys = set.size();
               } else {
                  I[i] = I[i-1] + prev_num_keys;
                  prev_num_keys = set.size();
               }
            });
         }  

         #pragma omp parallel for schedule(static, 1) num_threads(5)
         for (size_t i = 0; i < map.subcnt(); ++i) {
            map.with_submap(i, [&](const map_str_int_type::EmbeddedSet& set) {
               auto it = I[i];
               for (auto& x : set)
                  *it++ = std::make_pair(std::move(x.first), x.second);
            });
            map.reset_inner(i);             // <--- need ability to reset the inner sub map here
         }
```

I added the `reset_inner` method to `phmap.h`, locally on my machine.

```diff
diff --git a/parallel_hashmap/phmap.h b/parallel_hashmap/phmap.h
index d9a5b7b..c8a3d62 100644
--- a/parallel_hashmap/phmap.h
+++ b/parallel_hashmap/phmap.h
@@ -3444,6 +3444,10 @@ public:
         return  sets_[idx];
     }
 
+    void reset_inner(size_t idx) {
+        sets_[idx].set_ = EmbeddedSet();
+    }
+
     // Extension API: support for heterogeneous keys.
     //
     //   std::unordered_set<std::string> s;
```

Great news!

This change enables `llil4map` to reach `llil4hmap` performance, processing variable length keys. Previously, the `llil4map` demonstration ran slower. Being able to clear/reset a sub map early is beneficial when the sub map is no longer needed.  In the C++ snippet above, parallel threads process each sub map individually, "map to vector" below. I prefer for the map memory consumption to decrease while the vector memory increases and not wait until the end to clear the map. No issues for fixed-length keys.

**[llil4map.cc](https://gist.github.com/marioroy/862fa2fc6aa3b6f523f7a6ef9dd8d157)**

The sub maps are managed by the phmap library.

```text
$ ./llil4map in/big* in/big* in/big* | cksum
llil4map start
use OpenMP
use boost sort       Before 18.4GB     After 15.1GB
get properties         9.032 secs        8.969 secs
map to vector          3.054 secs        2.019 secs    <--- improvement
vector stable sort     3.099 secs        2.944 secs
write stdout           0.587 secs        0.553 secs
total time            15.774 secs       14.486 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

**[llil4hmap.cc](https://gist.github.com/marioroy/3924c48e140f8330f25f67cd98a815ef)**

The sub maps are managed at the application level. Notice, similar times for "get properties" and "map to vector". The key `hash_value` is computed once only and stored with the key. I was curious at the time and left it this way.

```text
$ ./llil4hmap in/big* in/big* in/big* | cksum
llil4hmap start
use OpenMP
use boost sort
get properties         8.896 secs
map to vector          2.075 secs
vector stable sort     2.957 secs
write stdout           0.586 secs
total time            14.516 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

Now, I understand why the two `phmap` demonstrations did not perform similarly before. You may like to know how far behind `emhash`. The `llil4map` is 0.9 seconds apart with `reset_inner` capability. Previously, 2.2 seconds apart total time. Please, no worries about phmap vs. emhash. Lacking the ability to clear/reset the sub map was the reason greater than 2 seconds apart (for variable length keys).

**[llil4emh.cc](https://gist.github.com/marioroy/7a69a422d88c0314f215e612ba607d2a)**

Here too, the sub maps are managed at the application level. Including, key `hash_value` stored with the key.

```text
$ ./llil4emh in/big* in/big* in/big* | cksum
llil4emh start
use OpenMP
use boost sort
get properties         7.767 secs
map to vector          2.192 secs
vector stable sort     3.075 secs
write stdout           0.543 secs
total time            13.578 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

**Comment by greg7mdp** [comment]
Thanks for the great explanation @marioroy .

Can you do the reset inside the `with_submap_n`, as in:

```c++
         for (size_t i = 0; i < map.subcnt(); ++i) {
            map.with_submap(i, [&](const map_str_int_type::EmbeddedSet& set) {
               auto it = I[i];
               for (auto& x : set)
                  *it++ = std::make_pair(std::move(x.first), x.second);
               set = map_str_int_type::EmbeddedSet();    // <------------------------------
            });
            map.reset_inner(i);             // <--- need ability to reset the inner sub map here
         }
```

**Comment by marioroy** [comment]
I made the change per your suggestion (removing `const`) and inserted line.
```c++
         for (size_t i = 0; i < map.subcnt(); ++i) {
            map.with_submap(i, [&](map_str_int_type::EmbeddedSet& set) {
               auto it = I[i];
               for (auto& x : set)
                  *it++ = std::make_pair(std::move(x.first), x.second);
               set = map_str_int_type::EmbeddedSet();
            });
         }
```

The `clang++` compiler fails.
```text
In file included from llil4map.cc:51:
./parallel-hashmap/parallel_hashmap/phmap.h:3431:9: error: no matching function for call to object of type '(lambda at llil4map.cc:415:32)'
 3431 |         fCallback(set);
      |         ^~~~~~~~~

llil4map.cc:415:17: note: in instantiation of function template specialization 'phmap::priv::parallel_hash_set<12, phmap::priv::raw_hash_set, spinlock_mutex, phmap::priv::FlatHashMapPolicy<std::basic_string<char>, unsigned int>, phmap::priv::StringHashEqT<char>::Hash, phmap::priv::StringHashEqT<char>::Eq, std::allocator<std::pair<const std::basic_string<char>, unsigned int>>>::with_submap<(lambda at llil4map.cc:415:32)>' requested here
  415 |             map.with_submap(i, [&](map_str_int_type::EmbeddedSet& set) {
      |                 ^

llil4map.cc:415:32: note: candidate function not viable: 1st argument ('const EmbeddedSet' (aka 'const raw_hash_set<phmap::priv::FlatHashMapPolicy<std::basic_string<char, std::char_traits<char>, std::allocator<char>>, unsigned int>, phmap::priv::StringHashEqT<char>::Hash, phmap::priv::StringHashEqT<char>::Eq, std::allocator<std::pair<const std::basic_string<char, std::char_traits<char>, std::allocator<char>>, unsigned int>>>')) would lose const qualifier
  415 |             map.with_submap(i, [&](map_str_int_type::EmbeddedSet& set) {
      |                                ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1 error generated.
```

**Comment by greg7mdp** [comment]
Sorry, my bad. 

After the loop, and instead if `map.reset_inner(i);`, use the following which is equivalent to your `reset_inner`:
```c++
map.with_submap_m(i, [&](map_str_int_type::EmbeddedSet& set) {
   set = map_str_int_type::EmbeddedSet();
});
```

**Comment by marioroy** [comment]
Thank you, @greg7mdp. Your suggestion works.
```text
         for (size_t i = 0; i < map.subcnt(); ++i) {
            map.with_submap(i, [&](const map_str_int_type::EmbeddedSet& set) {
               auto it = I[i];
               for (auto& x : set)
                  *it++ = std::make_pair(std::move(x.first), x.second);
            });
            // The sub map is no longer needed. Reset the set to reclaim memory.
            map.with_submap_m(i, [&](map_str_int_type::EmbeddedSet& set) {
               set = map_str_int_type::EmbeddedSet();
            });
         }
```

Notice "map to vector" completing in 2 seconds. This was taking 3 seconds, previously.
```text
$ ./llil4map in/big* in/big* in/big* | cksum
llil4map start
use OpenMP
use boost sort
get properties         8.946 secs
map to vector          2.018 secs    <-- Yeah! Thank you!
vector stable sort     2.960 secs
write stdout           0.589 secs
total time            14.514 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

**Comment by greg7mdp** [comment]
Hey @marioroy , I just tried your `lill4map.cc` myself. Would you mind if I added it to phmap as an example (with your attribution comment at the top of `lill4map.cc` of course).

Which parameters did you use for `gen-llil.pl`. I used the ones from the comment and I got:

```
❯ NUM_THREADS=3 ../ex_lill4map big1.txt big2.txt big3.txt >out.txt
llil4map (fixed string length=12) start
use OpenMP
don't use boost sort
get properties         0.695 secs
map to vector          0.093 secs
vector stable sort     1.316 secs
write stdout           0.061 secs
total time             2.166 secs
    count lines     10545600
    count unique    10368458
```

**Comment by marioroy** [comment]
No problem. [llil4map.cc](https://gist.github.com/marioroy/862fa2fc6aa3b6f523f7a6ef9dd8d157) is final.

Thank you for your help; (1) some time ago, found one-off error in chunking; (2) at the time, helped with move to vector; (3) and now, helped ensure minimum memory consumption. Blessings and grace.

That's it, just as in the comments for generating the input files.
```text
perl gen-llil.pl big1.txt 200 3 1
perl gen-llil.pl big1.txt 200 3 1
perl gen-llil.pl big1.txt 200 3 1
```

I found my script for making 92 input files + shuffled (requires 2.8 GB). On my machine (where llil4map.cc resides), `in` is a symbolic link to `/data/input`.  Change the path accordingly.

```bash
#!/bin/bash
# Script for generating input files for llil4map.cc.

mkdir -p /data/input
cp gen-llil.pl shuffle.pl /data/input
cd /data/input

# create 26 random files
for n in $(perl -le "print for 'aa'..'az'"); do
   perl gen-llil.pl big$n 200 3 1
   perl shuffle.pl big$n >1; mv 1 big$n
done &

# create 26 random files
for n in $(perl -le "print for 'ba'..'bz'"); do
   perl gen-llil.pl big$n 200 3 1
   perl shuffle.pl big$n >2; mv 2 big$n
done &

# create 26 random files
for n in $(perl -le "print for 'ca'..'cz'"); do
   perl gen-llil.pl big$n 200 3 1
   perl shuffle.pl big$n >3; mv 3 big$n
done &

# create 14 random files (total 92 files)
for n in $(perl -le "print for 'da'..'dn'"); do
   perl gen-llil.pl big$n 200 3 1
   perl shuffle.pl big$n >4; mv 4 big$n
done &

wait
```

The default mode in `llil4map.cc` is `fixed string length=12`.  Commenting out `line 124` becomes variable length keys (using std::basic_string). This requires 16 GB to process 92 input files if commented out. Otherwise, much less for fixed-length keys.
```c++
// #define MAX_STR_LEN_L (size_t) 12
```

**Comment by marioroy** [comment]
My Perl Monks friend [eyepopslikeamosquito](https://perlmonks.org/?node_id=176576) helped me learn C++.

I enjoy parallel processing. So naturally, I tried: (1) chunk IO in get properties using `spinlock_mutex`; (2) move to vector, this too is parallel; (3) parallel sort; (4) finally, parallel output but orderly.

**Comment by greg7mdp** [comment]
> Wow! Do not mind. [llil4map.cc](https://gist.github.com/marioroy/862fa2fc6aa3b6f523f7a6ef9dd8d157) is final.

Thanks, I'll add it as an example then, it is a great example!

**Comment by greg7mdp** [comment]
Added! Looks like you did an amazing job learning C++!

I saw a small perf. issue in `lill4map.cc`, where using the non-mutable `with_submap` does not allow to `std::move` the string out of the map. Could be an issue if long strings are used. 

[Here]( https://github.com/greg7mdp/parallel-hashmap/blob/effca243e650cca3fb0fb2f18e1da7d1b38d47ba/examples/lill4map.cc#L415-L419) is my updated version.

**Comment by marioroy** [comment]
Thanks. I learned also, reading `phmap_fwd_decl.h` and your examples.

I tried your updated version. Resetting the set is beneficial for long strings. Without it, the "map to vector" takes 3 seconds for long strings and peaks near 18.4 GB. 
```c++
         for (size_t i = 0; i < map.subcnt(); ++i) {
            map.with_submap_m(i, [&](map_str_int_type::EmbeddedSet& set) {
               auto it = I[i];
               for (auto& x : set)
                  *it++ = std::make_pair(std::move(const_cast<str_type&>(x.first)), x.second);
               // reset the set (no longer needed) to reclaim memory early
               set = map_str_int_type::EmbeddedSet();
            });
         }
```

I respect your decision if you prefer leaving the line commented out,  long strings mode (line 119).
```c++
// #define MAX_STR_LEN_L (size_t) 12
```
Or default to fixed-length, uncomment line.
```c++
#define MAX_STR_LEN_L (size_t) 12
```

Here too, I respect your decision leaving this 0 or 1 (line 46). I enabled it locally to validate parallel sort. Disabled takes 58.0 seconds for "vector stable sort" (long strings mode), or  21.5 seconds (fixed-length mode).
```c++
#define USE_BOOST_PARALLEL_SORT 1
```

Testing was done on a 32 core (64 logical threads), AMD Ryzen Threadripper 3970X machine, DDR4 memory 3600 MHz.

```text
$ ./llil4map in/big* in/big* in/big* | cksum
llil4map start
use OpenMP
use boost sort       without reset      with reset
get properties         9.035 secs        9.004 secs
map to vector          3.068 secs        2.010 secs    <--- reset is beneficial
vector stable sort     2.972 secs        2.988 secs
write stdout           0.639 secs        0.563 secs
total time            15.716 secs       14.567 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

**Comment by greg7mdp** [comment]
Thanks @marioroy , I reverted the defaults to the way you had them, and also added the `reset the set`. 
On my system `AMD® Ryzen 9 7950x 16-core processor × 32 with DDR5 memory`, I get the following, which is not the same as you:

```
~/gi/g/parallel-hashmap/build_gcc12_rel/llil_utils master *1 +1 !2 ?4 ❯ ./run_llil4map                                                                    19s
llil4map (fixed string length=12) start
use OpenMP
use boost sort
get properties        12.296 secs
map to vector          1.300 secs
vector stable sort     2.133 secs
write stdout           3.247 secs
total time            18.978 secs
    count lines     970195200
    count unique    200485780
3341100189 1811165988
~/gi/g/parallel-hashmap/build_gcc12_rel/llil_utils master *1 +1 !2 ?4 ❯ cat ./run_llil4map                                                                19s
#!/bin/bash

cd tmp
../../ex_llil4map big* big* big* | cksum
```

**Comment by marioroy** [comment]
That is great performance for the AMD 7950x CPU. The "write stdout" taking 3.2 seconds is likely the L3 size difference between the two CPUs. I am providing the following output, because the 7950x CPU is amazing. In other words, my system is unable to reach 2x faster total time for being 64 logical cores. Building with `clang++` runs faster for me.

**g++**

```text
$ g++ -o llil4map -std=c++20 -fopenmp -Wall -O3 llil4map.cc -I./parallel-hashmap
$ ./llil4map in/big* in/big* in/big* | cksum
llil4map (fixed string length=12) start
use OpenMP
use boost sort
get properties         7.496 secs
map to vector          0.965 secs
vector stable sort     1.740 secs
write stdout           0.657 secs
total time            10.859 secs
    count lines     970195200
    count unique    200483043
```

**clang++**

```text
$ clang++ -o llil4map -std=c++20 -fopenmp -Wall -O3 llil4map.cc -I./parallel-hashmap
$ ./llil4map in/big* in/big* in/big* | cksum
llil4map (fixed string length=12) start
use OpenMP
use boost sort
get properties         7.496 secs
map to vector          0.791 secs
vector stable sort     1.137 secs
write stdout           0.591 secs
total time            10.017 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

Last year, I tried custom memory allocation. I was trying to reduce peak memory consumption to no avail, because it greatly impacted "write stdout". Thanks for the reset suggestion. Memory consumption no longer peaks high.

**Comment by greg7mdp** [comment]
It is weird that I don't have the same result for `count unique` as you do.  Maybe a difference in gen_llil.pl?

**Comment by marioroy** [comment]
No, not weird at all. Because of the random generator script. This is normal for unique count to differ per machine or whenever refreshing the input files. If you ran `gen_files` again, very likely for unique count to differ.

Do you want to add to .gitignore, the tmp dir?
```text
/examples/llil_utils/tmp
```

**Comment by greg7mdp** [comment]
Sure, will do, even though in my case I generate the files in my build directory. 

I tried with `clang-16` and indeed it is also faster on my system, esp the `write stdout`.

```
lil4map (fixed string length=12) start
use OpenMP
use boost sort
get properties        12.873 secs
map to vector          1.370 secs
vector stable sort     1.862 secs
write stdout           1.177 secs
total time            17.284 secs
    count lines     970195200
    count unique    200478308

```

**Comment by marioroy** [comment]
> Sure, will do, even though in my case I generate the files in my build directory.

Oh, I assumed wrongly the location.

> I tried with clang-16 and indeed it is also faster on my system, esp the write stdout.

That is amazing.

I checked and you have the same defaults. Mindfulness for users and why default fixed-length is important? So the application can run with lesser memory consumption, no surprises on machines with 8 ~ 16 GB. The reset was the missing piece.
```c++
#define USE_BOOST_PARALLEL_SORT 1
#define MAX_STR_LEN_L (size_t) 12
```

Folks may run with a partial list, not all 92 files. For example just `a` i.e. `biga* biga* biga*`.

**Comment by greg7mdp** [comment]
I have a question about the data files. They contain words followed by a number. Am I right to assume this number is always positive?
I'll see if I can make `get_properties` a little bit faster, this is fun :-)

Also I added a couple minor simplifications in the code, see [this](https://github.com/greg7mdp/parallel-hashmap/commit/38dd6b52940937cae06191b7fb39ad0d6aeea4d2)

**Comment by greg7mdp** [comment]
I'll play with it some more later. It is an interesting piece of code for sure!

**Comment by marioroy** [comment]
Early in the project, I ran my Linux box with `transparent_hugepage=always`. So, just in case you find results at PerlMonks by me and wonder how is it possible. Above, my results were captured with `transparent_hugepage=madvise`. The `always` setting improves "map to vector".

```text
# Linux command-line transparent_hugepage=always
$ cat /sys/kernel/mm/transparent_hugepage/enabled
[always] madvise never

$ clang++ -o llil4map -std=c++20 -fopenmp -Wall -O3 llil4map.cc -I./parallel-hashmap
$ ./llil4map big* big* big* | cksum
llil4map (fixed string length=12) start
use OpenMP
use boost sort        madvise        always
get properties         7.496 secs     7.472 secs
map to vector          0.791 secs     0.475 secs  <--- improvement
vector stable sort     1.137 secs     1.106 secs
write stdout           0.591 secs     0.598 secs
total time            10.017 secs     9.653 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

I find it interesting that "get properties" performs similarly with `madvise` or `always`. That's amazing. Great job in the `phmap` library. The `always` setting makes a noticeable impact for the [llil4vec](https://perlmonks.org/?node_id=11149907) program, using `std::vector`. Does that mean the `std::vector` container is less efficient at memory management?

Caution: `llil4vec` appends to a vector and gobbles memory fast. Running with `big* big* big*` peaks 30 GB. It may cause slight swapping in a 32 GB VM/machine.

```bash
$ echo "madvise" | sudo tee /sys/kernel/mm/transparent_hugepage/enabled
$ echo "always" | sudo tee /sys/kernel/mm/transparent_hugepage/enabled
```
```text
$ clang++ -o llil4vec -std=c++20 -fopenmp -Wall -O3 llil4vec.cc
$ ./llil4vec big* big* big* | cksum
use OpenMP
use boost sort        madvise        always
get properties        12.016 secs     5.867 secs  <--- improvement
sort properties        5.891 secs     5.962 secs
vector reduce          1.521 secs     1.506 secs
vector stable sort     1.039 secs     1.005 secs
write stdout           0.637 secs     0.681 secs
total time            21.107 secs    15.018 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

I ran also, with 6x `big* big* big* big* big* big*` (552 total) to better understand the impact of `llil4vec`, and appreciation for the `phmap` library. Here, the `llil4vec` program requires minimum 64 GB in VM/machine, due to reaching 55 GB peak. For comparison, the `llil4map` program peaks less than 7 GB.

```text
$ ./llil4vec big* big* big* big* big* big* | cksum
llil4vec (fixed string length=12) start
use OpenMP
use boost sort        madvise        always
get properties        23.279 secs    12.028 secs  <--- improvement
sort properties       12.159 secs    12.232 secs
vector reduce          2.274 secs     2.255 secs
vector stable sort     1.017 secs     1.046 secs
write stdout           0.652 secs     0.670 secs
total time            39.383 secs    27.887 secs
    count lines     1940390400
    count unique    200483043
2511908988 1891299111

$ ./llil4map big* big* big* big* big* big* | cksum
llil4map (fixed string length=12) start
use OpenMP
use boost sort        madvise        always
get properties        14.694 secs    14.570 secs
map to vector          0.817 secs     0.506 secs  <--- improvement
vector stable sort     1.120 secs     1.135 secs
write stdout           0.651 secs     0.653 secs
total time            17.283 secs    16.866 secs
    count lines     1940390400
    count unique    200483043
2511908988 1891299111
```

The `phmap` container is awesome. `llil4map` can process hundreds of input files. Possibly, thousands :-)

**Comment by greg7mdp** [comment]
Hey @marioroy , just wondering, if your solution the fastest one that exists for this problem?

**Comment by marioroy** [comment]
> if your solution the fastest one that exists for this problem?

Hi, @greg7mdp I am not aware of anything faster. 

The [llil4emh](https://gist.github.com/marioroy/7a69a422d88c0314f215e612ba607d2a) `emhash` demonstration is the fastest for fixed-length keys. For long strings or variable length keys, `llil4map` and `llil4emh` perform similarly. The map demonstrations scale to many CPUs cores with ease. These are the fastest beyond 12 CPU threads.

There is the [llil4vec](https://perlmonks.com/?node_id=11149907) implementation. This appends to a vector, no maps. So, may consume a lot of memory, particularly repeating keys. It may run faster up to some number of cores, but does not scale well beyond that. If daring to run `llil4vec`, the max you can do for 32 GB is `big* big* big*` and only fixed-length=12 max. It peaks near 28 GB. So, be careful. It may cause swapping.

llil4map peaks less than 7 GB for 200 million keys, fixed-length=12.

I created a backup plan, if the number of unique keys were to exceed memory capacity :-) Run with `NUM_MAPS=128` maximum, default 32.

1. [llil4tkh](https://perlmonks.com/?node_id=11153432)  tkrzw::ShardDBM demonstration. The hash dbms are managed by the C++ library.
2. [llil4tkh2](https://perlmonks.com/?node_id=11153433)  tkrzw::HashDBM demonstration. The hash dbms are managed by the application.

**Comment by greg7mdp** [comment]
Wow, thanks for all the information! You seem to have spent a lot of time on this. Is this something you are doing for work, or just for fun?

**Comment by marioroy** [comment]
There is one more problem to solve. That is for long strings or variable length keys.

The std::string (std::basic_string) container is not memory efficient, unfortunately. In long strings mode, `llil4map` peaks 17.8 GB for the same input files. That's horrendous and more than doubles fixed-length mode. It is much worst for the `llil4vec` vector demonstration. That one peaks 58 GB, requiring a VM or machine with 64 GB.

**Comment by marioroy** [comment]
I made a memory efficient version [llil4map2.cc](https://gist.github.com/marioroy/63665aaa6a7c505a4cb20581fd408a58), using a vector of pointers, to the phmap key-value pairs. That met having to construct the phmap outside the block, and no reset. This results in sorting taking longer, due to CPU cache unfriendly. However, memory consumption is significantly less.

```text
200mil unique keys   llil4map   llil4map2
------------------   --------   --------
fixed-length         < 8.0 GB     6.0 GB
long-strings          17.8 GB    12.3 GB
```

**Comment by marioroy** [comment]
I made a memory efficient version [llil4map2](https://gist.github.com/marioroy/63665aaa6a7c505a4cb20581fd408a58), using a vector of pointers, to the phmap key-value pairs. That met having to construct the phmap outside the block, and no reset. This results in sorting taking longer, due to more CPU cache misses. However, memory consumption is significantly less.

```text
200mil unique keys   llil4map   llil4map2
------------------   --------   --------
fixed-length         < 8.0 GB     6.0 GB
long-strings          17.8 GB    12.3 GB
```

I removed the limit for the number of threads for better performance, sorting and output. This is helpful if RAM data is scattered, not cache friendly.

**Comment by greg7mdp** [comment]
> There is one more problem to resolve. That is for long strings or variable length keys.

Give me a couple days, I'll provide a solution for that (fixed size string which also supports long strings and is as memory efficient as possible)

**Comment by greg7mdp** [comment]
Hi @marioroy ,

I added a new type `string_cnt` which store a short string if its length is < 11 bytes, otherwise use a pointer, so the program works for any string length, but uses little memore for short string, and even for long strings it uses less space than `std::string`.

I also used a set instead of a map, storing the `string_cnt` directly, so that we know that it will use 16 bytes per entry for short strings. 

Some other small changes and cleanup, I hope you don't mind.

see [this](https://github.com/greg7mdp/parallel-hashmap/commit/5f26979baa15f289e17a1f75dc78de5faf39b22c)

**Comment by marioroy** [comment]
Hi @greg7mdp,

Wow! I gave it a spin on this machine. This peaks at 6.9 GB. The best thus far for supporting long strings, and automatic.

```text
$ ./llil4mapg in/big* in/big* in/big* | cksum
llil4map start
use OpenMP
use boost sort
get properties         8.152 secs
map to vector          0.885 secs
vector stable sort     2.764 secs
write stdout           0.696 secs
total time            12.499 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

At your convenience, please make one tiny change to allow parallel sort to run on all CPUs threads, i.e. no limit. Replace `std::min(nthds, 32)` with `nthds`.

```text
$ ./llil4mapg in/big* in/big* in/big* | cksum
llil4map start
use OpenMP
use boost sort
get properties         8.104 secs
map to vector          0.894 secs
vector stable sort     1.859 secs
write stdout           0.723 secs
total time            11.581 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

I will study `string_cnt`. The awesome part is not having to specify fixed-length or long strings. It's automatic.

**Comment by greg7mdp** [comment]
Hi @marioroy , I just made that update.

**Comment by marioroy** [comment]
I reached out to [eyepopslikeamosquito](https://perlmonks.com/?node_id=176576), to let him know. He triggered my curiosity in another [thread](https://perlmonks.com/?node_id=11148465), to attempt the [Long List is Long](https://perlmonks.org/?node_id=11147822) challenge.

Thank you @greg7mdp, for the `string_cnt` solution.

**Comment by marioroy** [comment]
I reached out to [eyepopslikeamosquito](https://perlmonks.com/?node_id=176576), to let him know. He triggered my curiosity in another [thread](https://perlmonks.com/?node_id=11148465), to making an attempt at the original [Long List is Long](https://perlmonks.org/?node_id=11147822) challenge. This became a great opportunity to learning OpenMP in C++.

Thank you @greg7mdp, for the `string_cnt` solution.

**Comment by marioroy** [comment]
> From `string_cnt`, "For larger strings, uses 16 bytes + strlen(string) + 1"

Recently, I came across a stack overflow [thread](https://stackoverflow.com/questions/10315041/meaning-of-acronym-sso-in-the-context-of-stdstring).  The top answer contains a string class involving a union.

```c++
class string {
public:
    // all 83 member functions
private:
    size_type m_size;
    union {
        class {
            // This is probably better designed as an array-like class
            std::unique_ptr<char[]> m_data;
            size_type m_capacity;
        } m_large;
        std::array<char, sizeof(m_large)> m_small;
    };
};
```

I will try a union in `string_cnt`. Ah, so that means `cnt` must come first in the structure. Is this not worth the effort? Curiosity wonders about much lower memory consumption will reduce :-)

```c++
struct string_cnt {
   uint32_t  cnt;
   union {
      char *    str;
      char      extra[4];
   };
   ...
};
```

**Comment by marioroy** [comment]
Never mind. A union will not work, because extra[3] is a boolean condition whether valid string pointer. For long strings, only 3 chars is wasted space. Have I misunderstood the clever `string_cnt` structure?

**Comment by greg7mdp** [comment]
Well, the current `string_cnt` has already the first 12 bytes functioning as a union equivalent to:
```
union {
   char *str;
   char immediate[12];
};
```

so the first 12 bytes contain either a pointer to a long string, or the immediate string.

There would be no saving if we made the `immediate[12]` and `immediate[8]` to be the same size as the pointer, as the whole struct including the count would be 12 bytes, but when stored in a `vector` or a `set` the next struct would be aligned to an 8 byte boundary, the the 4 bytes apparently saved would be left unused.

And yes, we need a way to know whether the struct stores a pointer or an immediate string. I use the last byte `extra[3]`, and when it is 0 (meaning immediate string), the zero can also serve as the null terminator of the string (since we don't store its size), so we can use the full 11 other bytes for useful characters.

**Comment by greg7mdp** [comment]
Pointers have to be aligned on a 8 byte boundary. So if you were to put the 4 byte `cnt` first, the next 4 bytes would be wasted because the union would have to start on an 8 byte boundary.

**Comment by greg7mdp** [comment]
I don't think you can save more space in the struct itself. But in the case of long string, if they are all the same size, you can use a custom allocator like `boost::pool_allocator`, but that would not work in the general case where the strings are of different sizes.

**Comment by marioroy** [comment]
Blessings and grace, @greg7mdp. Thank you for the extra clarity. The `string_cnt` struct is clever.

> Pointers have to be aligned on a 8 byte boundary.

Just earlier, I tried (for fun) putting `cnt` first and wondered about the assertions complaining. Now I understand.

What if `cnt` contained both flag (1 bit, indicating a pointer) and count (31 bits) using masking? Will that help remove the unused gap for long strings. Doing so may improve sorting performance.

**Comment by greg7mdp** [comment]
If you want the same struct to work for both long and short strings, as we do I think, it has to contain at least a pointer (8 bytes) + a count. 

If the count was never greater than 8, we could store it in the 3 lower bits of the pointer (which are always 0), so the pointer + count would fit in 8 bytes, so we could have either a 7 byte immediate string + count, or a pointer + count.

If the count can be greater than 8, then we need more than 8 bytes, so 16 bytes is the minimum our struct will occupy (for alignment), unless we are even more tricky by moving the pointers to an aligned address before dereferencing them.

But honestly, I don't see how you can save more than a few bytes, and then also you reduce the size of immediate strings that are supported.

**Comment by marioroy** [comment]
Please, disregard. So... _"Instead of extra[3], store the flag (1 bit) indicating whether using a pointer int the `cnt` variable. 31 bits is plenty big for count"_ will result in no savings?

**Comment by marioroy** [comment]
I understand. Wow! `string_cnt` is clever. My apologies for my ramblings.

**Comment by greg7mdp** [comment]
No problem, it is not obvious to understand if C++ is not your main language for sure.

**Comment by marioroy** [comment]
Oops, sorry. I met to write over. I understand. Wow! `string_cnt` is clever. My apologies for my ramblings.

```c++
   string_cnt(string_cnt&& o) noexcept {
      if (o.extra[3]) {
         str = o.str;
         o.str = nullptr;
         extra[3] = 1;
      } else {
         std::strcpy((char *)(&str), o.get());
         extra[3] = 0;
      }
      cnt = o.cnt;
   }
```

**Comment by greg7mdp** [comment]
This is a move constructor, so this is stealing the data from `o`. 
If `o` has a pointer to a long string, we can just copy the pointer, note that we have a pointer in `extra[3]`, and set `o.str` to null so `o`'s destructor will not free the memory.

If `o` contains a small string, we just copy it and mark that we have a small string as well.

**Comment by marioroy** [comment]
I understand your `string_cnt` fully. Thank you. The magic lies in the `set` method. The `strcpy` may span the two variables `str` and `extra`, union like.

```text
         auto len = std::strlen(s);
         if (len >= buffsz) {
            str = new char[len+1];
            std::strcpy(str, s);
            extra[3] = 1;
         } else {
            std::strcpy((char *)(&str), s);
            extra[3] = 0;
         }
```

**Comment by marioroy** [comment]
My favorite line `std::strcpy((char *)(&str), s);`, union like.

**Comment by marioroy** [comment]
My favorite line `std::strcpy((char *)(&str), s);`, union behavior `str` and `extra[3 chars max]`.  Thus, 11 chars max.

There's more cleverness :-)  The extra[3] is dual-purpose. A flag or null-terminating.

**Comment by marioroy** [comment]
I understand your `string_cnt` fully. Thank you. The magic lies in the `set` method. The `strcpy` may span the two variables `str` and `extra` up to 3 chars (4th char is a flag).

```text
         auto len = std::strlen(s);
         if (len >= buffsz) {
            str = new char[len+1];
            std::strcpy(str, s);
            extra[3] = 1;
         } else {
            std::strcpy((char *)(&str), s);
            extra[3] = 0;
         }
```

My favorite line std::strcpy((char *)(&str), s);, union behavior str and extra[3 chars max]. Thus, 11 chars max.

There's more cleverness :-) The extra[3] is dual-purpose. A flag or null-terminating.

**Comment by marioroy** [comment]
Your beautification of the include directives put me to work :-)  For consistency reason, I thought to beautify the include directives as well. I create a thread at Perl Monks, to let folks know: https://perlmonks.org/?node_id=11158711

I brought back [llil4umap](https://gist.github.com/marioroy/f1385f95e6a25119a3f7c1747f938222) (unordered_map variant). Your tip for releasing memory immediately resolved the long cleanup delay in that one. So, this is nice to have for comparing the std container vs the powerful phmap library.

**Comment by marioroy** [comment]
Your beautification of the include directives put me to work :-)  For consistency reason, I thought to beautify the include directives as well. I created a thread at Perl Monks to let folks know: https://perlmonks.org/?node_id=11158711

I brought back [llil4umap](https://gist.github.com/marioroy/f1385f95e6a25119a3f7c1747f938222) (unordered_map variant). Your tip for releasing memory immediately resolved the long cleanup delay in that one. So, the umap version is nice to have for comparing the standard container vs the powerful phmap library.

**Comment by marioroy** [comment]
@greg7mdp Your solution is superb. Thanks many, I learned a lot. The `string_cnt` struct boggles my mind on its clever design. My vote goes out to you.

Blessings and grace,

**Comment by marioroy** [comment]
@greg7mdp Your solution is superb. Thanks many, I learned a lot. The `string_cnt` struct boggles my mind on its clever design. You helped bring back sanity in my life. :-) Now, I no longer think about the Long List is Long challenge.

Blessings and grace,

**Comment by marioroy** [comment]
@greg7mdp You may have encountered the long global cleanup time, exiting the `llil4map` + `substring_cnt` program.

It dawned on me to do a testing involving mixed length. I created `long??` input files containing 6 chars or 15 chars. Basically, alternating `200 12 1` or `200 3 1` arguments to `gen-llil.pl`.

```bash
# create 26 random files
for n in $(perl -le "print for 'aa'..'az'"); do
   perl gen-llil.pl long$n 200 12 1
   perl shuffle.pl long$n >1; mv 1 long$n
done &

# create 26 random files
for n in $(perl -le "print for 'ba'..'bz'"); do
   perl gen-llil.pl long$n 200 3 1
   perl shuffle.pl long$n >2; mv 2 long$n
done &

# create 26 random files
for n in $(perl -le "print for 'ca'..'cz'"); do
   perl gen-llil.pl long$n 200 12 1
   perl shuffle.pl long$n >3; mv 3 long$n
done &

# create 14 random files (total 92 files)
for n in $(perl -le "print for 'da'..'dn'"); do
   perl gen-llil.pl long$n 200 3 1
   perl shuffle.pl long$n >4; mv 4 long$n
done &

wait
```

For this round of testing, the UNIX time is captured to see if any extra global/gc cleanup time when the program exits. That is the case for `substring_cnt`.

```text
1. llil4map_greg   - Greg's llil4map version with string_cnt
2. llil4map_long   - llil4map with the MAX_STR_LEN_L line commented out
3. llil4map_fix16  - llil4map with MAX_STR_LEN_L 16
```

**llil4map_greg**

```text
$ time ./llil4map_greg in/long* in/long* in/long* | cksum
llil4map start
use OpenMP
use boost sort
get properties        12.370 secs
map to vector          1.670 secs
vector stable sort     9.336 secs
write stdout           3.929 secs
total time            27.307 secs
    count lines     970195200
    count unique    295748977
3333632426 4307239210

real    0m43.433s
user    23m22.291s
sys     0m42.527s
```

**llil4map_long**

WARNING: Do not run llil4map_long (peaks 32 GB), or process  longa* longa* longa* (a suffix).

```text
$ time ./llil4map_long in/long* in/long* in/long* | cksum
llil4map start
use OpenMP
use boost sort
get properties        10.232 secs
map to vector          3.249 secs
vector stable sort     4.737 secs
write stdout           1.340 secs
total time            19.560 secs
    count lines     970195200
    count unique    295748977
3333632426 4307239210

real    0m20.269s
user    13m21.889s
sys     0m24.138s
```

**llil4map_fix16**

```text
$ time ./llil4map_fix16 in/long* in/long* in/long* | cksum
llil4map (fixed string length=16) start
use OpenMP
use boost sort
get properties         8.672 secs
map to vector          1.553 secs
vector stable sort     1.931 secs
write stdout           1.476 secs
total time            13.633 secs
    count lines     970195200
    count unique    295748977
3333632426 4307239210

real    0m13.785s
user    10m23.834s
sys     0m13.176s
```

**Comment by greg7mdp** [comment]
Interesting!

What happens is when you use `std::string`, the gnu standard library uses the same trick I did and does not allocate memory for strings up to 16 characters long (it is called the `short string optimization` or `sso`).

So in your test, with the mix of 6 character strings and 15 character string, you didn't allocate memory for strings whether you commented `MAX_STR_LEN_L` or not.

However, using `string_cnt`, that same optimization is limited to 11 bytes.

Probably if you tried with a mix 6/11 characters, or 6/17 characters, the difference would be less.

The aim with `string_cnt` was to reduce memory usage. indeed for string sizes between from 12 to 15 characters it will be slower.

**Comment by marioroy** [comment]
The speed is not the issue. No problem there. The interesting part is global cleanup. There is a 16 second gap from "total time" and the UNIX real time.

**Comment by marioroy** [comment]
> What happens is when you use std::string ...

I will re-create 19 length, alternating `200 16 1` or `200 3 1` arguments to gen-llil.pl. Well, to see if std::string does the same thing , long GC or global cleanup when the program exits. That was the idea of testing, check overall behavior of dynamic memory allocation.

**Comment by marioroy** [comment]
Wow! It std::string does the same time; long exit time due to GC or global cleanup. Thank you for the clarity about std::string sso. I wanted testing to be apples-to-apples comparison.

```text
$ time ./llil4map_greg in/long* in/long* in/long* | cksumllil4map start
use OpenMP
use boost sort
get properties        12.440 secs
map to vector          1.656 secs
vector stable sort     9.348 secs
write stdout           4.056 secs
total time            27.502 secs
    count lines     970195200
    count unique    295755152
29612263 5038456270

real    0m43.777s
user    23m27.969s
sys     0m43.282s
```

```text
$ time ./llil4map_long in/long* in/long* in/long* | cksum
llil4map start
use OpenMP
use boost sort
get properties        13.111 secs
map to vector          3.153 secs
vector stable sort    13.529 secs
write stdout           3.055 secs
total time            32.849 secs
    count lines     970195200
    count unique    295755152
29612263 5038456270

real    0m49.815s
user    20m59.538s
sys     0m59.482s
```

Your program wins, consuming lesser memory. I will run again and edit this comment with the peak memory consumption.

**Comment by marioroy** [comment]
Fixed length capability is helpful for mitigating the exit global cleanup time. The UNIX time reduced by ~ 3x. Here, I'm running llil4map with MAX_STR_LEN_L 20 (19 + 1). Peak memory consumption: 18.2 GB

**llil4map_fix20**

```text
$ time ./llil4map_fix20 in/long* in/long* in/long* | cksum
llil4map (fixed string length=20) start
use OpenMP
use boost sort
get properties         9.000 secs
map to vector          1.810 secs
vector stable sort     3.076 secs
write stdout           1.923 secs
total time            15.810 secs
    count lines     970195200
    count unique    295755152
29612263 5038456270

real    0m15.994s
user    11m21.552s
sys     0m15.691s
```

**Comment by marioroy** [comment]
std::string and string_cnt, once pass sso limit, result in significantly longer exit time due to global cleanup. This challenge has been interesting.

**Comment by marioroy** [comment]
string_cnt :-)  Allow one to specify the fix-length limit during construction. To have `MAX_STR_LEN_L`-like capability.

1. A key size of 11 chars is quite small.
2. Once past sso limit, not fun, affects overall program completion time significantly
3. A solution to this madness then, is dynamic fix-length capability up to some point ~ 30 max

**Comment by marioroy** [comment]
@greg7mdp Akin to { sha1sum, sha224sum, sha256sum, sha384sum, sha512sum }, one can build llil4map supporting fixed-mode llil4map_12, llil4map_16, llil4map_20, llil4map_24, llil4map_30, and llil4map_dyn (using static_cnt).

The Chuma LLiL challenge is solved. :-)

**Comment by greg7mdp** [comment]
> If only `string_cnt` allowed one to specify the fix-length limit, for `MAX_STR_LEN_L`-like capability.

It is indeed possible. I have previously modified the version of  `string_cnt` to use `std::string_view`, but here is a version which is templatized to support various `sso` sizes:

```
// ---------------------------------------------------------------------------------------------
// Stores a string + a count
// For strings up to N-1 bytes, total space used is N + 4 bytes
// For larger strings, uses 16 bytes + strlen(string) + 1
//
// invariants
//    if  extra[mark_idx], str is a valid string pointer
//    if !extra[mark_idx], the N bytes starting at (const char *)(&str) store a null_terminated string
// ---------------------------------------------------------------------------------------------
template<size_t N>
struct string_cnt_t {
   using uint_t = uint32_t;

   static_assert(N >= 12);
   static constexpr size_t extra_sz = N - sizeof(char*);
   static constexpr size_t mark_idx = extra_sz - 1;

   char *   str;
   char     extra[extra_sz];
   uint_t   cnt;

   static constexpr size_t buffsz = sizeof(str) + sizeof(extra);

   string_cnt_t() : str{nullptr}, extra{0}, cnt{0} {}

   string_cnt_t(std::string_view s, uint_t c = 0) : str(nullptr), cnt(c) { set(s); }

   ~string_cnt_t() { free(); }

   string_cnt_t(const string_cnt_t& o) {
      set(o.get());
   }

   string_cnt_t(string_cnt_t&& o) noexcept {
      if (o.extra[mark_idx]) {
         str = o.str;
         o.str = nullptr;
         extra[mark_idx] = 1;
      } else {
         std::strcpy((char *)(&str), o.get());
         extra[mark_idx] = 0;
      }
      cnt = o.cnt;
   }

   string_cnt_t& operator=(const string_cnt_t& o) {
      free();
      set(o.get());
      cnt = o.cnt;
      return *this;
   }

   string_cnt_t& operator=(string_cnt_t&& o) noexcept {
      free();
      new (this) string_cnt_t(std::move(o));
      return *this;
   }

   std::strong_ordering operator<=>(const string_cnt_t& o) const { return std::strcmp(get(), o.get()) <=> 0; }
   bool operator==(const string_cnt_t& o) const { return std::strcmp(get(), o.get()) == 0; }

   std::size_t hash() const {
      auto s = get();
      std::string_view sv {s};
      return std::hash<std::string_view>()(sv);
   }

   const char *get() const { return extra[mark_idx] ? str : (const char *)(&str); }

private:
   void free() { if (extra[mark_idx]) { delete [] str; str = nullptr; } }

   void set(std::string_view s) {
      static_assert(buffsz == 12);
      static_assert(offsetof(string_cnt_t, cnt) == (intptr_t)buffsz);
      static_assert(sizeof(string_cnt_t) == 16);

      assert(!extra[mark_idx] || !str);
      if (s.empty())
         std::memset(&str, 0, buffsz);
      else {
         auto len = s.size();
         if (len >= buffsz) {
            str = new char[len+1];
            std::memcpy(str, s.data(), len);
            str[len] = 0;
            extra[mark_idx] = 1;
         } else {
            std::memcpy((char *)(&str), s.data(), len);
            ((char *)&str)[len] = 0;
            extra[mark_idx] = 0;
         }
      }
   }

   void set(const char *s) {
      set(std::string_view{s});
   }
};

using string_cnt = string_cnt_t<12>;

namespace std {
   template<> struct hash<string_cnt> {
      std::size_t operator()(const string_cnt& v) const noexcept { return v.hash(); };
   };
}
```

**Comment by greg7mdp** [comment]
PS: I got `get_properties` to be faster... takes 8.8s on my machine with the 6 character strings.

**Comment by marioroy** [comment]
The new update is wonderful.

I like the fact that the private method set no longer calls strlen. Because, the length of the key is already known in get_properties. Previously (version 1), lazy_emplace_l accepted beg_str. So now, we give it a string_view variable? It compiled without errors or warnings.

```c++
            int_type count = fast_atoll64(found + 1);
            std::string_view sv(beg_ptr, found - beg_ptr);

            // Use lazy_emplace to modify the set while the mutex is locked.
            set_ret.lazy_emplace_l(
               sv,
               [&](string_cnt_set_t::value_type& p) {
                  p.cnt += count; // called only when key was already present
               },
               [&](const string_cnt_set_t::constructor& ctor) {
                  ctor(std::move(sv), count); // construct value_type in place when key not present
               }
            );
```

Here too, **get_properties is faster** (strlen(s), strcpy vs. s.size(), memcpy()). Here are my before and after results.

```text
$ ./llil4mapg big* big* big* | cksum
llil4map start
use OpenMP
use boost sort
get properties         8.104 secs     7.763 secs
map to vector          0.894 secs     0.937 secs
vector stable sort     1.859 secs     1.867 secs
write stdout           0.723 secs     0.674 secs
total time            11.581 secs    11.243 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

I re-compiled llil4map_greg.cc using `string_cnt_t<20>`. **Two asserts failed**.

```c++
   void set(std::string_view s) {
      static_assert(buffsz == 12);
      static_assert(offsetof(string_cnt_t, cnt) == (intptr_t)buffsz);
      static_assert(sizeof(string_cnt_t) == 16);
```

I changed two assert lines.

```c++
   void set(std::string_view s) {
      static_assert(buffsz == N);
      static_assert(offsetof(string_cnt_t, cnt) == (intptr_t)buffsz);
      static_assert(sizeof(string_cnt_t) == N + 4);
```

**Let's try fixed-length 20**. Before and after.

```text
$ time ./llil4map_greg long* long* long* | cksum
llil4map start
use OpenMP
use boost sort
get properties        12.440 secs     9.270 secs   <-- improvement
map to vector          1.656 secs     2.118 secs   humm, we can bump thds_move
vector stable sort     9.348 secs     3.294 secs   <-- improvement
write stdout           4.056 secs     1.668 secs   <-- improvement
total time            27.502 secs    16.352 secs
    count lines     970195200
    count unique    295755152
29612263 5038456270

            real    0m43.777s      0m16.823s   <-- no GC side effect
            user   23m27.969s     13m35.370s
            sys     0m43.282s      0m15.187s
```

**I bumped `nthds_move` to 12** (previously, 6). No need to go higher. Let's try again :-).

```text
   int nthds_move = std::min(nthds, 12);
```

```text
$ time ./llil4map_greg long* long* long* | cksum
llil4map start
use OpenMP
use boost sort
get properties        12.440 secs     9.277 secs   <-- improvement
map to vector          1.656 secs     1.853 secs   <-- very good
vector stable sort     9.348 secs     3.293 secs   <-- improvement
write stdout           4.056 secs     1.626 secs   <-- improvement
total time            27.502 secs    16.051 secs
    count lines     970195200
    count unique    295755152
29612263 5038456270

            real    0m43.777s      0m16.517s   <-- no GC side effect
            user   23m27.969s     13m35.582s
            sys     0m43.282s      0m15.290s
```

Wow! **The improved `string_cnt` provides flexibility**, allowing the developer to choose the default fixed-length size. Also, no problem if the input has long strings sparingly, exceeding sso. It will still work, with minimum impact to GC cleanup.

At last, **the LLiL challenge is solved**. Thanks to you, @greg7mdp. The `string_cnt` update is amazing.

**Comment by marioroy** [comment]
The new update is wonderful.

I like the fact that the private method `set` no longer calls `strlen`. Because, the length of the key is already known in get_properties. Previously (version 1), lazy_emplace_l accepted beg_str. So now, we give it a string_view variable? It compiled without errors or warnings.

```c++
            int_type count = fast_atoll64(found + 1);
            std::string_view sv(beg_ptr, found - beg_ptr);

            // Use lazy_emplace to modify the set while the mutex is locked.
            set_ret.lazy_emplace_l(
               sv,
               [&](string_cnt_set_t::value_type& p) {
                  p.cnt += count; // called only when key was already present
               },
               [&](const string_cnt_set_t::constructor& ctor) {
                  ctor(std::move(sv), count); // construct value_type in place when key not present
               }
            );
```

Here too, get_properties is faster (strlen(s), strcpy vs. s.size(), memcpy()). Here are my before and after results.

```text
$ ./llil4mapg big* big* big* | cksum
llil4map start
use OpenMP
use boost sort
get properties         8.104 secs     7.763 secs
map to vector          0.894 secs     0.937 secs
vector stable sort     1.859 secs     1.867 secs
write stdout           0.723 secs     0.674 secs
total time            11.581 secs    11.243 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

I re-compiled llil4map_greg.cc using `string_cnt_t<20>`. **Two asserts failed**.

```c++
   void set(std::string_view s) {
      static_assert(buffsz == 12);
      static_assert(offsetof(string_cnt_t, cnt) == (intptr_t)buffsz);
      static_assert(sizeof(string_cnt_t) == 16);
```

Fixing, I changed the first and third assert lines.

```c++
   void set(std::string_view s) {
      static_assert(buffsz == N);
      static_assert(offsetof(string_cnt_t, cnt) == (intptr_t)buffsz);
      static_assert(sizeof(string_cnt_t) == N + 4);
```

Let's try fixed-length 20. Before and after results.

```text
$ time ./llil4map_greg long* long* long* | cksum
llil4map start
use OpenMP
use boost sort
get properties        12.440 secs     9.270 secs   improved
map to vector          1.656 secs     2.118 secs   humm, we can bump thds_move
vector stable sort     9.348 secs     3.294 secs   improved
write stdout           4.056 secs     1.668 secs   improved
total time            27.502 secs    16.352 secs
    count lines     970195200
    count unique    295755152
29612263 5038456270

            real    0m43.777s      0m16.823s   improved, no GC side effect
            user   23m27.969s     13m35.370s
            sys     0m43.282s      0m15.187s
```

**I bumped `nthds_move` to 12** (previously, 6). No need to go higher. Let's try again :-).

```text
   int nthds_move = std::min(nthds, 12);
```

```text
$ time ./llil4map_greg long* long* long* | cksum
llil4map start
use OpenMP
use boost sort
get properties        12.440 secs     9.277 secs   improved
map to vector          1.656 secs     1.853 secs   very good
vector stable sort     9.348 secs     3.293 secs   improved
write stdout           4.056 secs     1.626 secs   improved
total time            27.502 secs    16.051 secs
    count lines     970195200
    count unique    295755152
29612263 5038456270

            real    0m43.777s      0m16.517s   improved, no GC side effect
            user   23m27.969s     13m35.582s
            sys     0m43.282s      0m15.290s
```

Wow! The improved `string_cnt` provides flexibility, allowing the developer to specify the sso fixed-length size. Also, no problem if the input has long strings sparingly, exceeding sso. It will still work, with minimum impact to GC cleanup.

At last, the LLiL challenge is solved. Thanks to you, @greg7mdp. The `string_cnt` update is amazing.

P.S. Not fair :-)  I too wish to use `string_cnt` in my phmap examples `#include <parallel_hashmap/string_cnt.h>` or `#include <examples/string_cnt.h>`.

**Comment by greg7mdp** [comment]
That's great, thanks for all your work. And for fixing the static_assert in `string_cnt`. I still have some work to do on my new version but I'm not sure when I'll get it done.

> I too want to use string_cnt in my phmap demonstrations 

You are welcome to use `string_cnt` for whatever you want.

**Comment by marioroy** [comment]
So much learning. I sped up get_properties by storing the length of the key with the count i.e. (count << 8) + (klen & 0xff). 

1. I replaced two set lines from `set(o.get());` to this `set(o.get(), o.cnt & 0xff);` and changed the private set method.

```c++
   void set(const char *s, uint8_t len) {
      set(std::string_view{s, len});
   }
```

2. The hash method is called hundreds of million times. The 1 second improvement is from here.

```c++
   std::size_t hash() const {
      auto s = get();
      std::string_view sv {s, cnt & 0xff};
      return std::hash<std::string_view>()(sv);
   }
```

I ran with 16 CPU cores. Before and after.

```text
$ NUM_THREADS=16 taskset -c 0-15 ./llil4map_greg in/big* in/big* in/big* | cksum
llil4map start
use OpenMP
use boost sort
get properties        23.146 secs     22.106 secs  <-- improved
map to vector          0.818 secs      0.846 secs
vector stable sort     4.547 secs      4.496 secs
write stdout           0.538 secs      0.535 secs
total time            29.049 secs     27.985 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

get_properties is now as fast as my llil4map version (without string_cnt).

**Comment by greg7mdp** [comment]
Very nice speedup! congrats.

**Comment by marioroy** [comment]
I created a gist (two files) containing your `llil4map_greg.cc` and `string_cnt.h`. Modification: Count contains key length.

https://gist.github.com/marioroy/0f97def75739eff7ed9fee7071da3629

**Comment by marioroy** [comment]
Fixed assert typo in set (replaced 4 with extra_sz). Correction applied to gist and my reply above.
```c++
      static_assert(sizeof(string_cnt_t) == N + extra_sz);
```

**Comment by marioroy** [comment]
I commented out the assert, complaining for `string_cnt_t<20>`.

Last testing!  I compared std::string sso(16) and string_cnt sso(16). In my version `llil4map.cc`, I commented out the `MAX_STR_LEN_L` definition. In `llil4map_greg.cc`, using `string_cnt_t<16>`.  That's as close to apples-to-apples comparison. The third column is `string_cnt<20>` for comparison.

```text
$ ./llil4map big* big* big* | cksum
llil4map start       std::string     string_cnt<16>
use OpenMP           18.3 GB peak    10.9 GB peak
use boost sort       ------------    ------------
get properties         9.030 secs      8.130 secs
map to vector          2.199 secs      1.288 secs
vector stable sort     3.062 secs      2.051 secs
write stdout           0.799 secs      0.843 secs
total time            15.093 secs     12.313 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689

$ time ./llil4map long* long* long* | cksum
llil4map start       std::string     string_cnt<16>  string_cnt<20>
use OpenMP           36.7 GB peak    24.7 GB peak    19.3 GB peak
use boost sort       ------------    ------------    ------------
get properties        13.192 secs     12.757 secs      9.380 secs
map to vector          3.178 secs      1.873 secs      1.896 secs
vector stable sort    13.151 secs      9.852 secs      3.304 secs
write stdout           3.234 secs      3.868 secs      1.580 secs
total time            32.757 secs     28.351 secs     16.162 secs
    count lines     970195200
    count unique    295755152
29612263 5038456270

 wall clock real    0m49.593s       0m45.068s       0m16.620s
            user   20m55.395s      24m15.965s      13m41.070s
            sys     0m56.593s       0m46.258s       0m14.461s
```

This is a good closure with the Long List is Long (LLiL) challenge. During the journey, `std::string` felt like a curse, gobbling memory consumption. Thank you for `string_cnt`. There are no more challenges :-) What a relief :-)

**Comment by marioroy** [comment]
Sorting an array runs faster for fixed length keys. That is the case for `block_indirect_sort`. So, I tried union std::array in the struct (see `string_cnt_u.h` at [gist](https://gist.github.com/marioroy/0f97def75739eff7ed9fee7071da3629)). Note: The gist contains both header files.

```c++
   union {
      struct {
         char *   str;
         char     extra[extra_sz];
         uint_t   cnt;
      } s_;
      std::array<char, sizeof(s_) - sizeof(uint_t)> d_;
   };

   bool operator<(const string_cnt_t& o) const {
      if (s_.extra[mark_idx] || o.s_.extra[mark_idx])
         return std::strcmp(getstr(), o.getstr()) < 0;
      else
         return d_ < o.d_;
   }
```

Before and after results.

```text
$ ./llil4map_greg big* big* big* | cksum
llil4map start
use OpenMP
use boost sort        string_cnt.h  string_cnt_u.h
get properties         7.763 secs     7.689 secs
map to vector          0.937 secs     0.836 secs
vector stable sort     1.867 secs     1.589 secs  <-- improved (64 threads)
write stdout           0.674 secs     0.719 secs
total time            11.243 secs    10.835 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

Temporarily, I decreased the number of threads to 8 for sorting. Look at sorting go. :-)

```text
$ ./llil4map_greg in/big* in/big* in/big* | cksumllil4map start
use OpenMP
use boost sort        string_cnt.h  string_cnt_u.h
get properties         7.703 secs     7.682 secs
map to vector          0.900 secs     0.898 secs
vector stable sort     9.264 secs     7.283 secs  <-- improved (8 threads)
write stdout           0.763 secs     0.640 secs
total time            18.631 secs    16.506 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

Gees! All the union changes just to improve sorting for fixed-length keys.

**Comment by greg7mdp** [comment]
> ```c++
> return d_ < o.d_;
> ```

Hum, this comparison is not necessarily correct, if you have garbage after the terminating null character of the contained string, because the `std::array` `<` will not stop at the null character.

**Comment by greg7mdp** [comment]
If you `set()` a long string, the first 8 bytes contain a pointer. Now if you set a 2 byte string "ab", the first 8 bytes will contain "ab\0?????". I imagine this doesn't happen in llil.

**Comment by marioroy** [comment]
No problem. The str is initialized to null pointer. So, all nulls in the array union. The comparison using `d_` only if both this "and" other are fixed length (no pointers).

**Comment by marioroy** [comment]
No problem. The str is initialized to null pointer. So, all nulls in the array union. The less than comparison does strcmp if either this "or" other has a pointer. That is, array comparison only when this "and" other are fixed length, no pointers.
```c++
   bool operator<(const string_cnt_t& o) const {
      if (s_.extra[mark_idx] || o.s_.extra[mark_idx])
         return std::strcmp(getstr(), o.getstr()) < 0;
      else
         return d_ < o.d_;
   }
```

**Comment by marioroy** [comment]
I tried a small string. No garbage, instead trailing nulls (as described above). Btw, gist update: Added missing include `<cstdint>` to header files.

```c++
// clang++ -o test -std=c++20 -Wall -O3 test.cc

#include "string_cnt_u.h"
#include <iostream>

int main(void) {
    string_cnt_t<12> s("hi", 2);
    for (int i = 0; i < 12; i++)
        std::cout << (int) s.d_[i] << '\n';
    return 0;
}
```

```text
$ ./test 
104
105
0
0
0
0
0
0
0
0
0
0
```

**Comment by marioroy** [comment]
I tried a small string. No garbage, instead trailing nulls (as described above). Btw, gist update: Added missing include `<cstdint>` to header files.

```c++
// clang++ -o test -std=c++20 -Wall -O3 test.cc

#include "string_cnt_u.h"
#include <iostream>

int main(void) {
    string_cnt_t<12> s("hi", 2);
    for (int i = 0; i < 12; i++)
        std::cout << (int) s.d_[i] << '\n';
    return 0;
}
```

```text
$ ./test 
104
105
0
0
0
0
0
0
0
0
0
0
```

Blank string `string_cnt_t<12> s("", 0);` No garbage.

```text
0
0
0
0
0
0
0
0
0
0
0
0
```

**Comment by marioroy** [comment]
I replaced the `strcpy` line with memcpy. This benefits sorting significantly. Note: I renamed the `d_` variable to `f_`, (indicating fixed length union array).

```c++
   string_cnt_t(string_cnt_t&& o) noexcept {
      if (o.s_.extra[mark_idx]) {
         s_.str = o.s_.str;
         o.s_.str = nullptr;
         s_.extra[mark_idx] = 1;
      } else {
         // std::strcpy((char *)(&s_.str), o.getstr());
         // s_.extra[mark_idx] = 0;
         std::memcpy(f_.data(), o.f_.data(), N);
      }
      s_.cnt = o.s_.cnt;
   }
```

**string_cnt_t<20>**

For long fixed lengths, string_cnt_u surpasses my version. It turns out that `f_ < o.f_` is faster than `std::memcmp(f_.data(), o.f_.data(), N) < 0`. My version does memcmp.

```text
$ ./llil4map long* long* long* | cksum
llil4map (fixed string length=20) start
use OpenMP                           strcpy str    memcpy array
use boost sort       mario's 4map   string_cnt_u   string_cnt_u
get properties         9.042 secs     9.208 secs     9.298 secs
map to vector          1.888 secs     1.860 secs     1.890 secs
vector stable sort     9.312 secs    12.861 secs     7.583 secs  <- 8 thds
write stdout           1.603 secs     1.630 secs     1.622 secs
total time            21.846 secs    25.560 secs    20.448 secs
    count lines     970195200
    count unique    295755152
29612263 5038456270
```

**string_cnt_t<12>**

For small fix lengths, `f_ < o.f_` (4.786 secs, total time 13.972 secs) is slower than `std::memcmp(f_.data(), o.f_.data(), N) <=> 0` (3.639 secs, total time 12.904 secs). Below, using memcmp for sorting for apples-to-apples comparison with the old version.

```text
$ ./llil4map big* big* big* | cksum
llil4map (fixed string length=12) start
use OpenMP                           strcpy str    memcpy array
use boost sort       mario's 4map   string_cnt_u   string_cnt_u
get properties         7.561 secs     7.682 secs     7.673 secs
map to vector          0.889 secs     0.898 secs     0.888 secs
vector stable sort     3.125 secs     7.283 secs     3.639 secs  <- 8 thds
write stdout           0.684 secs     0.640 secs     0.703 secs
total time            12.260 secs    16.506 secs    12.904 secs
    count lines     970195200
    count unique    200483043
2057246516 1811140689
```

The last two days, I learned about `std::array < std::array`, which may run slower or faster than `memcmp` depending on the size. The great news is the `llil4map_greg.cc` variant is now fast.

@greg7mdp I'm finished testing. The gist has both `string_cnt.h` and `string_cnt_u.h`.

**Comment by marioroy** [comment]
Hi @greg7mdp,

You were correct. There could be garbage, fixed by clearing in set. Thank you!

```c++
            std::memset(f_.data(), 0, N);
            std::memcpy(f_.data(), s.data(), std::min(len, N - 1));
```

Fixing the garbage issue allowed me to improve the `==` operator in `string_cnt_u.h`.

```c++
   bool operator==(const string_cnt_t& o) const {
      if (s_.extra[mark_idx] || o.s_.extra[mark_idx])
         return std::strcmp(getstr(), o.getstr()) == 0;
      else
         return std::memcmp(getdata(), o.getdata(), N) == 0;
   }
```

Freeing memory is now handled with multiple threads, `llil4map_greg.cc`.

```c++
   // Free dynamically allocated memory in parallel.
   #pragma omp parallel for schedule(dynamic, 300000) num_threads(nthds)
   for (size_t i = 0; i < propvec.size(); i++)
      propvec[i].free();
```

I added a fix-length only version, `string_cnt_f.h`.

```c++
// #include "string_cnt.h"     // original + key length stored with cnt
// #include "string_cnt_f.h"   // fix-length only version
   #include "string_cnt_u.h"   // union std::array
```

Thank you for `string_cnt.h`. The `llil4map_greg.cc` program runs fast (`string_cnt_u.h`, slightly faster `string_cnt_f.h`).

**Comment by greg7mdp** [comment]
Hi @marioroy , congrats on all the great work. I've been very busy at work, so I haven't had time to work on this or look at your latest changes, but hopefully I will soon. All the best!

**Comment by marioroy** [comment]
Greetings, @greg7mdp

I made a new gist containing my final phmap and vector demonstrations. The gist contains a README and LLiL utilities. There are three `string_cnt_t` variants; with extra comments.

C++ phmap and vector demonstrations using string_cnt_t
https://gist.github.com/marioroy/693d952b578792bf090fe20c2aaccad5

1. `llil4map_next.cc` - phmap::parallel_flat_hash_map demonstration
2. `llil4vec_next.cc` - similarly, a std::vector demonstration
3. `string_cnt.h` - original + key length stored with count
4. `string_cnt_f.h` - fixed-length-only version
5. `string_cnt_u.h` - union std::array
6. `gen-files-big.sh` - generate 92 big files (total 2.8 GB)
7. `gen-files-long.sh` - generate 92 long files (total 5.0 GB)
8. `gen-llil.pl` - generate LLiL test file
9. `shuffle.pl` - shuffle LLiL test file

I learned more C++ from your `string_cnt.h` example, so much appreciated. The LLiL challenge is solved. `llil4map_next.cc` runs with low memory consumption and pleasantly fast using `string_cnt_u.h`.

**Comment by marioroy** [comment]
Hi @greg7mdp, `get_properties` processing `long* long* long*` took 16.9 seconds previously, now 14.1 seconds.

```c++
            int_type count = fast_atoll64(found + 1);
            string_cnt s{(const char *)beg_ptr, (size_t)(found - beg_ptr), count};

            // Use lazy_emplace to modify the set while the mutex is locked.
            set_ret.lazy_emplace_l(
               s,
               [&](string_cnt_set_t::value_type& p) {
                  p.incr(count);      // called only when key was already present
               },
               [&](const string_cnt_set_t::constructor& ctor) {
                  ctor(std::move(s)); // construct value_type in place when key not present
               }
            );
```

**Comment by marioroy** [comment]
I made an attempt at dynamic buffer allocation. Get properties decreased to 10.6s on a realtime Linux, now similar to running on non-realtime kernel.

https://gist.github.com/marioroy/d02881b96b20fa1adde4388b3e216163

The dynamic allocation logic is thread-safe i.e. first argument thread id, optional. Plus rolling back the dynamic memory position if the long string was already present.

```c++
   int_type count = fast_atoll64(found + 1);
   string_cnt s{tid, beg_ptr, (size_t)(found - beg_ptr), count};

   // Use lazy_emplace to modify the set while the mutex is locked.
   set_ret.lazy_emplace_l(
      s,
      [&](string_cnt_set_t::value_type& p) {
         p.incr(count); // called only when key was already present
         s.mem_rollback(tid);
      },
      [&](const string_cnt_set_t::constructor& ctor) {
         // construct value_type in place when key not present
         // long strings will be stored in dynamic memory storage
         ctor(std::move(s));
      }
   );
```

Notice lesser memory consumption. And there too, get properties completing faster than before.

```text
$ time ./llil4map long* long* long* | cksum
llil4map start
use OpenMP             ( old )    ( next )     ( buf )
memory consumption     36.6 GB     17.5 GB     16.0 GB
use boost sort
get properties         12.967s     11.527s     10.402s
map to vector           3.244s      1.252s      1.211s
vector stable sort     10.265s      8.708s      8.161s
write stdout            1.410s      1.388s      1.386s
total time             27.887s     22.877s     21.163s
    count lines      970195200
    count unique     295755152
29612263 5038456270

       real time     0m43.973s   0m24.872s   0m21.909s
       user         25m08.532s  22m58.120s  20m56.492s
       sys           0m55.602s   0m48.783s   0m11.551s
```

Thank you, @greg7mdp. Your `string_cnt_t` struct is how I learned to do this.
